basics/data-import/schema-conforming-transformer.md

```diff
@@ -0,0 +1,101 @@
+---
+description: Storing records with dynamic schemas in a table with a fixed schema.
+---
+
+# Ingest records with dynamic schemas
+
+Some domains (e.g., logging) generate records where each record can have a different set of keys, whereas Pinot tables have a relatively static schema. For records with varying keys, it's impractical to store each field in its own table column. However, most (if not all) fields may be important, so fields should not be dropped unnecessarily.
+
+The [SchemaConformingTransformer](https://github.com/apache/pinot/blob/master/pinot-segment-local/src/main/java/org/apache/pinot/segment/local/recordtransformer/SchemaConformingTransformer.java) is a [RecordTransformer](https://github.com/apache/pinot/blob/master/pinot-segment-local/src/main/java/org/apache/pinot/segment/local/recordtransformer/RecordTransformer.java) that can transform records with dynamic schemas such that they can be ingested in a table with a static schema. The transformer primarily takes record fields that don't exist in the schema and stores them in a type of catchall field.
+
+For example, consider this record:
+
+```json
+{
+  "timestamp": 1687786535928,
+  "hostname": "host1",
+  "HOSTNAME": "host1",
+  "level": "INFO",
+  "message": "Started processing job1",
+  "tags": {
+    "platform": "data",
+    "service": "serializer",
+    "params": {
+      "queueLength": 5,
+      "timeout": 299,
+      "userData_noIndex": {
+        "nth": 99
+      }
+    }
+  }
+}
+```
+
+Let's say the table's schema contains the following fields:
+* timestamp
+* hostname
+* level
+* message
+* tags.platform
+* tags.service
+* indexableExtras
+* unindexableExtras
+
+Without this transformer, the `HOSTNAME` field and the entire `tags` field would be dropped when storing the record in the table. However, with this transformer, the record would be transformed into the following:
+
+```json
+{
+  "timestamp": 1687786535928,
+  "hostname": "host1",
+  "level": "INFO",
+  "message": "Started processing job1",
+  "tags.platform": "data",
+  "tags.service": "serializer",
+  "indexableExtras": {
+    "tags": {
+      "params": {
+        "queueLength": 5,
+        "timeout": 299
+      }
+    }
+  },
+  "unindexableExtras": {
+    "tags": {
+      "userData_noIndex": {
+        "nth": 99
+      }
+    }
+  }
+}
+```
+
+Notice that the transformer does the following:
+* Flattens nested fields which exist in the schema, like `tags.platform`
+* Drops some fields like `HOSTNAME`, where `HOSTNAME` must be listed as a field in the config option `fieldPathsToDrop`
+* Moves fields which don't exist in the schema and have the suffix `_noIndex` into the `unindexableExtras` field (the field name is configurable)
+* Moves any remaining fields which don't exist in the schema into the `indexableExtras` field (the field name is configurable)
```

Comment:
```suggestion
* Moves fields that don't exist in the schema and have the suffix `_noIndex` into the `unindexableExtras` field (the field name is configurable)
* Moves any remaining fields that don't exist in the schema into the `indexableExtras` field (the field name is configurable)
```

Reply:
does "the field name is configurable" mean that the suffix could be "_noIndex" or be configured to be something else? or does it mean that "userData" could be any field? if it means the latter, think we can drop as understood by this being an example

Reply:
I meant that the destination field could have a different name than "indexableExtras" (e.g., "extras"). But I agree, it's confusing which one we're referring to, so I dropped it.